core/store/src/trie/mem/updating.rs

```diff
@@ -33,15 +71,100 @@ pub enum UpdatedMemTrieNode {
     },
     Extension {
         extension: Box<[u8]>,
-        child: OldOrUpdatedNodeId,
+        child: GenericNodeOrIndex<GenericTrieNodePtr>,
     },
     /// Corresponds to either a Branch or BranchWithValue node.
     Branch {
-        children: Box<[Option<OldOrUpdatedNodeId>; 16]>,
+        children: Box<[Option<GenericNodeOrIndex<GenericTrieNodePtr>>; 16]>,
         value: Option<FlatStateValue>,
     },
 }
 
+pub type UpdatedMemTrieNode = GenericUpdatedTrieNode<MemTrieNodeId, FlatStateValue>;
+
+/// An updated node with its memory usage.
+/// Needed to recompute subtree function (memory usage) on the fly.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct GenericUpdatedTrieNodeWithSize<GenericTrieNodePtr, FlatStateValue> {
+    pub node: GenericUpdatedTrieNode<GenericTrieNodePtr, FlatStateValue>,
+    pub memory_usage: u64,
+}
+
+impl<N, V> GenericUpdatedTrieNodeWithSize<N, V> {
+    pub fn empty() -> Self {
+        Self { node: GenericUpdatedTrieNode::Empty, memory_usage: 0 }
+    }
+}
+
+impl<GenericTrieNodePtr, FlatStateValue: HasLength>
+    GenericUpdatedTrieNode<GenericTrieNodePtr, FlatStateValue>
+{
+    fn memory_usage_value(value_length: u64) -> u64 {
+        value_length * TRIE_COSTS.byte_of_value + TRIE_COSTS.node_cost
+    }
+
+    pub fn memory_usage_direct(&self) -> u64 {
+        match self {
+            Self::Empty => {
+                // DEVNOTE: empty nodes don't exist in storage.
+                // In the in-memory implementation Some(TrieNode::Empty) and None are interchangeable as
+                // children of branch nodes which means cost has to be 0
+                0
+            }
+            Self::Leaf { extension, value } => {
+                TRIE_COSTS.node_cost
+                    + (extension.len() as u64) * TRIE_COSTS.byte_of_key
+                    + Self::memory_usage_value(value.len())
+            }
+            Self::Branch { value, .. } => {
+                TRIE_COSTS.node_cost
+                    + value.as_ref().map_or(0, |value| Self::memory_usage_value(value.len()))
+            }
+            Self::Extension { extension, .. } => {
+                TRIE_COSTS.node_cost + (extension.len() as u64) * TRIE_COSTS.byte_of_key
+            }
+        }
+    }
+}
+
+pub type UpdatedMemTrieNodeWithSize = GenericUpdatedTrieNodeWithSize<MemTrieNodeId, FlatStateValue>;
```

Comment:
Confirming.... It seems like we don't use the `size` parameter for MemTrieNodes anywhere? Could you please confirm this? In which case we can remove this type in favor of `GenericUpdatedTrieNode` without size?

Reply:
I don't see how to do it without adding comparable overhead.
The `GenericUpdatedTrieNode` serves as a generic enum. Adding size to it is the same as implementing `GenericUpdatedTrieNodeWithSize`.
If I look from `GenericTrieUpdate` perspective, both node and memory has to be taken simultaneously if `generic_get_node` is called, and "node with size" struct seems to be useful here again.
Could you be more specific on how to get rid of `GenericUpdatedTrieNodeWithSize`?

Reply:
Let me explore a bit myself on this and then suggest

Reply:
Aah crap, you're right! I was under the impression that memory calculation happens independently at the end of update/insert function, but seems like it's built into the function call. I suppose what we have currently is best even though I believe it's irrelevant for memtries? Or have I got that detail wrong?

Reply:
Spoke offline